### PR TITLE
feat(cheatcodes): Make `expectCall` only work for the next call's subcalls

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

Right now, `expectCall` works at the "test" level—as long as there are enough calls to the target contract at any depth level in the test, it will pass. This will make it so that the calls are only counted if they happen in an external call, meaning calls performed at the "test" ("root") level will not count.

## New behavior

The current behavior allows using expectCall this way:

```solidity
// PASS
function testThings() {
  Contract inner = new Contract();
  NestedContract target = new NestedContract(inner);
  vm.expectCall(address(inner), abi.encodeWithSelector(inner.bar.selector), 2);
  inner.bar(); // call bar directly. Counts as 1 of two calls expected.
  target.baz(); // will call bar internally. This counts as 2 of two calls expected. expectCall is fulfilled.
}
```

This is not the intended behavior, due to this reason:
- `expect*` cheatcodes should only work for the next call. 
 
This means that `expectCall` should actually expect to match calls that are "subcalls" of the next call.

This PR introduces the following behavior

```solidity
// FAIL
function testOne() {
  Contract inner = new Contract();
  NestedContract target = new NestedContract(inner);
  vm.expectCall(address(inner), abi.encodeWithSelector(inner.bar.selector), 1);
  inner.bar(); // call bar directly. This does not get counted, as this is a test-level call. It should be abstracted into another call.
}

// PASS
function testOne() {
  Contract inner = new Contract();
  NestedContract target = new NestedContract(inner);
  vm.expectCall(address(inner), abi.encodeWithSelector(inner.bar.selector), 1);
  target.baz(); // will call bar internally, therefore works.
}

contract NestedContract {
  Contract public inner;

  constructor(Contract _inner) {
    inner = _inner;
  }
  function baz() public {
    inner.bar(); // call bar 
  }
} 
```

TODO:
- [ ] Book update

### CODE REVIEW

**Comment by Evalir** [comment]
the ci failure looks a tad strange @mattsse — this shouldn't be making revm randomly fail to check the journaled state? we're not accessing it on places it might be undefined i believe, but it seems like the usage of `depth` in the `call` hook might be making it fail?

**Review by mattsse — COMMENTED** [review]
seems like only the transact_fork test is affected, 
could also be an rpc provider issue with that test, but it's weird that this panics

File: evm/src/executor/inspector/cheatcodes/mod.rs [hunk 1]
```diff
@@ -884,7 +891,10 @@ where
                     }
                 }
             }
+        }
 
+        // If the depth is 0, then this is the root call terminating
+        if data.journaled_state.depth() == 0 {
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/mod.rs:897]: 👍

**Review by mattsse — APPROVED** [review]
nice, 
tests lgtm!

**Comment by Evalir** [comment]
Reverting this to merge into a feature branch

**Comment by PaulRBerg** [comment]
Could you cut another release please, @mattsse?

Our tests are failing now:

https://app.warp.dev/block/mLxbng37Ztgs9OxSko3PMf

**Pull request merged**